### PR TITLE
[#168454509] Reliability improvements

### PR DESCRIPTION
**Pull request by AP-Hunt**
What
---
Whilst investigating why DIT's metric exporter was crashing, I spotted a couple of partially-related problems. This PR contains a number of commits aimed at generally improving the quality and reliability of paas-prometheus-exporter

How to review
---
A code review will likely be sufficient.

Who can review
---
Anyone

### CODE REVIEW

File: cf/client.go [hunk 1]
```diff
@@ -196,3 +198,26 @@ func (l *logCacheHTTPClient) Do(req *http.Request) (*http.Response, error) {
...
+	)
+
+	for i = 0; i < maxRetries; i++ {
+		token, err = tokenSource.Token()
+
+		if err == nil {
+			return token, err
+		}
+
+		log.Println(fmt.Sprintf("getting token failed (attempt %d of %d). Retrying. Error: %s", i, maxRetries, err.Error()))
```
Comment (philandstuff) [inline cf/client.go]: Why not `log.Printf`? Why `log.Println` and `fmt.Sprintf`?

File: service/watcher.go [hunk 2]
```diff
@@ -154,3 +155,33 @@ func (w *Watcher) Close() {
...
+	)
+
+	for i = 0; i < maxRetries; i++ {
+		envelopes, err = client.Read(ctx, serviceGuid, start)
+
+		if err == nil {
+			return envelopes, err
+		}
+
+		log.Println(fmt.Sprintf("reading log-cache lines for service %s failed (attempt %d of %d). Error: %s", serviceGuid, i, maxRetries, err.Error()))
```
Comment (philandstuff) [inline service/watcher.go]: ditto why not `log.Printf`?

File: main.go [hunk 3]
```diff
@@ -123,7 +123,7 @@ func main() {
 
 	for {
 		select {
-		case <-errChan:
+		case err := <-errChan:
```
Comment (tlwr) [inline main.go]: Oooh finally resolved!

File: cf/client.go [hunk 4]
```diff
@@ -196,3 +198,26 @@ func (l *logCacheHTTPClient) Do(req *http.Request) (*http.Response, error) {
...
+	var (
+		i     int
+		token *oauth2.Token
+		err   error
+	)
+
+	for i = 0; i < maxRetries; i++ {
+		token, err = tokenSource.Token()
+
+		if err == nil {
```
[the remark this answers is not in the document]
Comment (AP-Hunt) [reply cf/client.go]: Yeah that's a fair shout. It's a little inside out compared to the convention.

File: cf/client.go (hunk 1 above)
Comment (AP-Hunt) [reply cf/client.go]: Good question. Don't think I was looking too much when I wrote this!

File: cf/client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (AP-Hunt) [reply cf/client.go]: Addressed

File: cf/client.go (hunk 1 above)
Comment (AP-Hunt) [reply cf/client.go]: Addressed

**Comment by philandstuff** [comment]
one team one dream ❤️

**Comment by AP-Hunt** [comment]
Still more to come on this one :O

**Comment by philandstuff** [comment]
> Still more to come on this one :O

😕 why wouldn't you want to merge what you currently have and open a new PR for the rest? It looks great and ready for prime time

**Comment by AP-Hunt** [comment]
> > Still more to come on this one :O
> 
> 😕 why wouldn't you want to merge what you currently have and open a new PR for the rest? It looks great and ready for prime time

Because Lyndon at DIT is deploying from this branch in his staging env to give the changes a proper test. There's actually a larger story here around improving the reliability of the thing so it doesn't crash as often.

**Pull request merged**